Implement a --cubical=no-glue option

yeejian-tan opened the pull request:
This pull request aims to implement a `--cubical-without-glue` pragma option that allows one to use Cubical features, but disallowing the primitive Glue type. This is in anticipation of my ongoing project to implement Cubical with UIP as mentioned in #3750 .

The key changes made were
- addition of the `--cubical-without-glue` flag with tests
- ensuring no importation of Glue into files with `--cubical-without-glue` flag by a custom `InfectiveImport` record
- a new Cubical variant via the `CWithoutGlue :: Cubical` datatype constructor
- many modules with `--erased-cubical` which do not use Glue in erased positions are changed to `--cubical-without-glue` instead

A question I have in mind: now that we will have 3 different cubical options (not counting `--cubical-compatible`): Full, Erased, and WithoutGlue, perhaps it is more ergonomic/intuitive to unify the command line flags (`--cubical`, `--erased-cubical`, `--cubical-without-glue`) that set the same variable into `--cubical=[full,erased,noglue]` instead (with `--cubical` meaning `--cubical=full` by default)?

This is my first time contributing to Agda, please let me know if there is anything I missed. Many thanks in advance!

yeejian-tan committed:
Implement --cubical-without-glue

yeejian-tan committed:
tests for --cubical-without-glue

yeejian-tan committed:
update test outputs to include new flag

andreasabel commented on test/Fail/Impossible.err:
```diff
@@ -2,3 +2,4 @@ ret > ExitFailure 154
 out > An internal error has occurred. Please report this as a bug.
 out > Location of the error: __IMPOSSIBLE__, called at src/full/Agda/ImpossibleTest.hs:«line»:«col» in «Agda-package»:Agda.ImpossibleTest
 out >   impossibleTest, called at src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs:«line»:«col» in «Agda-package»:Agda.Syntax.Translation.ConcreteToAbstract
+out >
```
Undo this change please, it causes CI to fail.

andreasabel commented on test/Fail/ImpossibleVerbose.err:
```diff
@@ -3,3 +3,4 @@ out > This message should be debug-printed with option '-v impossible:10'.
 out > An internal error has occurred. Please report this as a bug.
 out > Location of the error: __IMPOSSIBLE_VERBOSE__, called at src/full/Agda/ImpossibleTest.hs:«line»:«col» in «Agda-package»:Agda.ImpossibleTest
 out >   impossibleTest, called at src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs:«line»:«col» in «Agda-package»:Agda.Syntax.Translation.ConcreteToAbstract
+out >
```
Undo this change please, it causes CI to fail.

andreasabel commented on test/Fail/ImpossibleVerboseReduceM.err:
```diff
@@ -3,3 +3,4 @@ out > ReduceM should also print this debug message.
 out > An internal error has occurred. Please report this as a bug.
 out > Location of the error: __IMPOSSIBLE_VERBOSE__, called at src/full/Agda/ImpossibleTest.hs:«line»:«col» in «Agda-package»:Agda.ImpossibleTest
 out >   impossibleTestReduceM, called at src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs:«line»:«col» in «Agda-package»:Agda.Syntax.Translation.ConcreteToAbstract
+out >
```
Undo this change please, it causes CI to fail.

andreasabel commented on test/Fail/CubicalWithoutGlueCannotPrimGlue.agda:
```diff
@@ -0,0 +1,11 @@
+primitive
+  primGlue    : ∀ {ℓ ℓ'} (A : Set ℓ) {φ : I}
+        → (T : Partial φ (Set ℓ')) → (e : PartialP φ (λ o → T o ≃ A))
+        → Set ℓ'
```
Formatting error.

andreasabel reviewed (commented):
I'd approve; but merging has to wait until after the 2.8.0 release

nad commented:
Documentation is missing.

Is the idea that it should be possible to import modules that use `--cubical-without-glue` from modules that use `--erased-cubical` or `--cubical`?

nad commented:
> This is in anticipation of my ongoing project to implement Cubical with UIP as mentioned in #3750 .

Please discuss the design of such a feature before you implement it.

plt-amy commented:
It would be good to have a decision on #7835 before adding yet another module-level variant of cubical agda.

yeejian-tan committed:
documentation for --cubical-without-glue

yeejian-tan commented:
> Documentation is missing.

I have just added some minimal documentation (for the [flag](https://agda--7861.org.readthedocs.build/en/7861/tools/command-line-options.html#cmdoption-cubical-without-glue) and on the [Cubical](https://agda--7861.org.readthedocs.build/en/7861/language/cubical.html) page), let me know if you think we need more.

> Is the idea that it should be possible to import modules that use `--cubical-without-glue` from modules that use `--erased-cubical` or `--cubical`?

Yes, that is what I have in mind.

nad commented:
I think the documentation should include something about how the options interact (which is [somewhat complicated](https://agda.readthedocs.io/en/v2.7.0.1/language/cubical.html#cubical-agda-with-erased-glue) in the case of `--erased-cubical` and `--cubical`).

yeejian-tan commented:
> I think the documentation should include something about how the options interact (which is [somewhat complicated](https://agda.readthedocs.io/en/v2.7.0.1/language/cubical.html#cubical-agda-with-erased-glue) in the case of `--erased-cubical` and `--cubical`).

https://github.com/agda/agda/pull/7861/commits/59e6946b61c80fed516ef55bc9213b07c6852594 adds more detailed documentation [as a subsection](https://agda--7861.org.readthedocs.build/en/7861/language/cubical.html#variants). I also included a table on how they interact, please let me know if there is any mistake:

![image](https://github.com/user-attachments/assets/f34f5c64-8bf9-450e-babe-9ecdbd880425)

andreasabel commented:
> `--cubical=[full,erased,noglue]`

`no-glue` please.

Should `--cubical-compatible` be writeable as `--cubical=compatible` as well?  I'd think so.

andreasabel commented:
@Mergifyio rebase

yeejian-tan committed:
Change Cubical flags to take string arguments

The flag --cubical now takes a string as input.
Empty string `--cubical` or `--cubical=full` represents full Cubical,
`--cubical=compatible` aliases `--cubical-compatible`,
`--cubical=erased` aliases `--erased-cubical`,
and `--cubical=no-glue` replaces the previously implemented `--cubical-without-glue`.

yeejian-tan committed:
fixup! documentation for --cubical-without-glue

yeejian-tan committed:
fixup! Change Cubical flags to take string arguments

yeejian-tan committed:
fixup! Change Cubical flags to take string arguments

yeejian-tan commented:
This is now ready for another round of review.

The technical changes stay the same, while the `--cubical-without-glue` flag is renamed into `--cubical=no-glue`, along with the aliasing of the other cubical-related flags (`--cubical[=full]`, `--cubical=compatible`, and `--cubical=erased`). The changes are summarised in `CHANGELOG.md`, put here for easy reference:

* added a `--cubical=no-glue` variant
* Consolidated Cubical-related flags to
  | Old                    | New                                               |
  | ---------------------- | ------------------------------------------------- |
  | `--cubical`            | `--cubical`, or `--cubical=full`                  |
  | `--erased-cubical`     | `--erased-cubical`, or `--cubical=erased`         |
  | `--cubical-compatible` | `--cubical-compatible`, or `--cubical=compatible` |
  | -                      | `--cubical=no-glue`                               |

The [user manual](https://agda--7861.org.readthedocs.build/en/7861/) is also updated.

I kept all the old flags for backwards compatibility, but new flags are shown in documentation, code comment, and error messages to "prepare" the users for the shift. We can schedule the deprecation for another release if we wanted to.

yeejian-tan committed:
Merge branch 'master' into no-glue

yeejian-tan committed:
fixup! Merge branch 'master' into no-glue

andreasabel commented:
Agda dev meeting 2025-10-15: sorry for the delay in handling this.  This PR has proponents and no opponents, so we agree to merge it (if review process is successful). @jespercockx will handle it.

jespercockx reviewed (approved):
I looked through this PR and everything looks good for it to be merged!

jespercockx committed:
Merge branch 'master' into SwampertX-no-glue

jespercockx commented:
I brought this back up to date with current master (doing a merge since it's anyway the plan to squash this PR).

(Edit: one advantage of doing a merge commit is that it makes Github prevent me from accidentally rebasing this PR instead of squashing it.)

jespercockx commented:
Let's merge this now and see what people have to say about it before the release of 2.9!

andreasabel commented on test/Fail/CubicalWithoutGlueCannotPrimGlue.agda:
```diff
@@ -0,0 +1,11 @@
+primitive
+  primGlue : ∀ {ℓ ℓ'} (A : Set ℓ) {φ : I}
+             → (T : Partial φ (Set ℓ')) → (e : PartialP φ (λ o → T o ≃ A))
+             → Set ℓ'
```
This test is flaky; I am getting this diff when I run the test locally:
```diff
-CubicalWithoutGlueCannotPrimGlue.agda:9.3-11.22: error: [NeedOptionCubical]
-Option --cubical[=full] required
-when checking that the type of the primitive function primGlue is
-∀ {ℓ} {ℓ'} (A : Set ℓ) {φ : I} →
-(T : Partial φ (Set ℓ')) →
-(e : PartialP φ (λ o → T o ≃ A)) → Set ℓ'
+CubicalWithoutGlueCannotPrimGlue.agda:5.1-39: error: [InfectiveImport]
+Importing module Agda.Builtin.Cubical.Equiv which might contain
+glue to a module with the option --cubical=no-glue.
+when scope checking the declaration
+  open import Agda.Builtin.Cubical.Equiv
```
I suspect it behaves differently when `.agdai` files are present already (local run) as opposed to when they are all created by the testsuite (CI run).

The explanation that I can think of is that this PR did not bump the interface version in `Agda.TypeChecking.Serialize`.

andreasabel replied:
Since there are new PRs that did bump the interface version, I was puzzled why I still get this problem.
Now it seems that a run of `agda --setup` fixed my problem, so this was a false alarm.

yeejian-tan replied:
Oops, I indeed forgot to bump. Thanks for resolving it in the end!
